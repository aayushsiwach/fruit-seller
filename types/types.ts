// types.ts
export interface ItemType {
    id: string;
    name: string;
    price: number;
    quantity: number;
    image: string;
    description: string;
    category: string;
    discount: number;
    is_seasonal: boolean;
    created_at: string;
}

export interface CartItem {
    product_id: string;
    quantity: number;
}

export interface Order {
    id: string;
    userName: string;
    user_id: string;
    items: CartItem[];
    total: number;
    createdAt: string;
    status: string;
}

export interface CartContextType {
    cart: CartItem[];
    loading: boolean;
    addToCart: (product: ItemType, quantity?: number) => void;
    updateQuantity: (productId: string, quantity: number, maxQuantity: number) => void;
    removeFromCart: (productId: string) => void;
    clearCart: () => void;
    getCartTotal: (products: (ItemType | undefined)[]) => number;
    getCartItemCount: () => number;
    showSnackbar: (message: string, severity: "success" | "error") => void; // Add for Snackbar
}

export interface LayoutProps {
    children: React.ReactNode;
}