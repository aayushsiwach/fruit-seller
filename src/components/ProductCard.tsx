import {
    Card,
    CardMedia,
    CardContent,
    CardActions,
    Typography,
    Button,
    Box,
    IconButton,
    TextField,
} from "@mui/material";
import { useRouter } from "next/router";
import { useCart } from "../contexts/CartContext";
import { FiShoppingCart, FiPlus, FiMinus } from "react-icons/fi";
import { ItemType } from "@/types/types";
import { motion } from "framer-motion";

const ProductCard = ({ product }: { product: ItemType }) => {
    const router = useRouter();
    const { cart, addToCart, updateQuantity, removeFromCart } = useCart();
    const cartItem = cart.find((item) => item.product_id === product.id);
    const cartQuantity = cartItem ? cartItem.quantity : 0;

    // Debug logs
    // console.log("ProductCard - Product ID:", product.id);
    // console.log("ProductCard - Cart Item:", cartItem);
    // console.log("ProductCard - Cart Quantity:", cartQuantity);

    const discountedPrice = product.discount
        ? product.price * (1 - product.discount / 100)
        : null;

    const isOutOfStock = product.quantity === 0;

    const handleAddToCart = (e: React.MouseEvent) => {
        e.stopPropagation();
        addToCart(product, 1);
    };

    const handleQuantityChange = (e: React.MouseEvent, newQuantity: number) => {
        e.stopPropagation();
        if (newQuantity <= 0) {
            removeFromCart(product.id);
        } else if (newQuantity <= product.quantity) {
            updateQuantity(product.id, newQuantity, product.quantity);
        }
    };

    const handleViewDetails = () => {
        router.push(`/products/${product.id}`);
    };

    return (
        <Card
            sx={{
                height: 370,
                width: "100%",
                display: "flex",
                flexDirection: "column",
                justifyContent: "space-between",
                transition: "transform 0.2s, box-shadow 0.2s",
                // opacity: isOutOfStock ? 0.6 : 1,
                // filter: isOutOfStock ? "grayscale(50%)" : "none",
                "&:hover": {
                    transform: isOutOfStock ? "none" : "translateY(-4px)",
                    boxShadow: isOutOfStock
                        ? "none"
                        : "0 12px 20px rgba(0, 0, 0, 0.1)",
                },
                cursor: isOutOfStock ? "default" : "pointer",
                position: "relative",
            }}
            onClick={handleViewDetails}
            component={motion.div}
            whileHover={{ scale: isOutOfStock ? 1 : 1.02 }}
            transition={{ duration: 0.2 }}
        >
            <CardMedia
                component="img"
                height="160"
                image={product.image || "/placeholder-fruit.jpg"}
                alt={product.name}
                sx={{ objectFit: "cover" }}
            />

            <CardContent sx={{ flexGrow: 1 }}>
                <Box
                    sx={{
                        display: "flex",
                        justifyContent: "space-between",
                        alignItems: "center",
                    }}
                >
                    <Typography
                        variant="h6"
                        fontWeight={600}
                        noWrap
                        title={product.name} // tooltip on hover
                        sx={{
                            overflow: "hidden",
                            textOverflow: "ellipsis",
                            whiteSpace: "nowrap",
                        }}
                    >
                        {product.name}
                    </Typography>
                    {isOutOfStock && (
                        <Typography
                            variant="body2"
                            color="error"
                            sx={{ mt: 1, fontWeight: 500 }}
                        >
                            Out of Stock
                        </Typography>
                    )}
                </Box>

                {discountedPrice ? (
                    <Box
                        sx={{
                            display: "flex",
                            flexDirection: "column",
                            // gap: 1,
                            alignItems: "start",
                        }}
                    >
                        <Typography variant="h6" color="green" fontWeight={700}>
                            Rs. {discountedPrice.toFixed(2)}
                        </Typography>
                        <Typography
                            variant="body2"
                            color="text.secondary"
                            sx={{ textDecoration: "line-through" }}
                        >
                            Rs. {product.price.toFixed(2)}
                        </Typography>
                    </Box>
                ) : (
                    <Typography variant="h6" fontWeight={700}>
                        Rs. {product.price.toFixed(2)}
                    </Typography>
                )}

                <Typography
                    variant="body2"
                    color="text.secondary"
                    mt={1}
                    title={product.description}
                    sx={{
                        display: "-webkit-box",
                        WebkitLineClamp: 2,
                        WebkitBoxOrient: "vertical",
                        overflow: "hidden",
                        minHeight: 40, // ensures space even if short
                    }}
                >
                    {product.description}
                </Typography>
            </CardContent>

            <CardActions sx={{ p: 2, pt: 0 }}>
                {cartQuantity > 0 ? (
                    <Box
                        sx={{
                            display: "flex",
                            alignItems: "center",
                            justifyContent: "space-between",
                            width: "100%",
                            gap: 1,
                        }}
                        component={motion.div}
                        initial={{ opacity: 0 }}
                        animate={{ opacity: 1 }}
                        transition={{ duration: 0.3 }}
                    >
                        <IconButton
                            size="small"
                            onClick={(e) =>
                                handleQuantityChange(e, cartQuantity - 1)
                            }
                            disabled={isOutOfStock}
                            component={motion.div}
                            whileHover={{ scale: 1.1 }}
                        >
                            <FiMinus />
                        </IconButton>
                        <TextField
                            value={cartQuantity}
                            onChange={(e) => {
                                e.stopPropagation();
                                const value = Number.parseInt(e.target.value);
                                if (
                                    !isNaN(value) &&
                                    value >= 0 &&
                                    value <= product.quantity
                                ) {
                                    if (value === 0) {
                                        removeFromCart(product.id);
                                    } else {
                                        updateQuantity(
                                            product.id,
                                            value,
                                            product.quantity
                                        );
                                    }
                                }
                            }}
                            inputProps={{
                                min: 0,
                                max: product.quantity,
                                style: { textAlign: "center" },
                            }}
                            sx={{ width: 50 }}
                            size="small"
                        />
                        <IconButton
                            size="small"
                            onClick={(e) =>
                                handleQuantityChange(e, cartQuantity + 1)
                            }
                            disabled={cartQuantity >= product.quantity}
                            component={motion.div}
                            whileHover={{ scale: 1.1 }}
                        >
                            <FiPlus />
                        </IconButton>
                    </Box>
                ) : (
                    <Button
                        variant="contained"
                        color="primary"
                        fullWidth
                        startIcon={<FiShoppingCart />}
                        onClick={handleAddToCart}
                        disabled={isOutOfStock}
                        component={motion.div}
                        whileHover={{ scale: 1.05 }}
                        whileTap={{ scale: 0.95 }}
                    >
                        Add to Cart
                    </Button>
                )}
            </CardActions>
        </Card>
    );
};

export default ProductCard;
