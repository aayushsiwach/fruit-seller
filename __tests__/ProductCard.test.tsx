import React from "react";
import { render } from "@testing-library/react";
import ProductCard from "@/src/components/ProductCard";
// import { useCart } from "@/src/contexts/CartContext";
import { ItemType } from "@/types/types";

jest.mock("next/router", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
}));

jest.mock("@/src/contexts/CartContext", () => ({
  useCart: () => ({
    cart: [],
    addToCart: jest.fn(),
    updateQuantity: jest.fn(),
    removeFromCart: jest.fn(),
  }),
}));

const sampleProduct: ItemType = {
  id: "1",
  name: "Grapes",
  price: 100,
  discount: 10,
  quantity: 20,
  image: "/grapes.jpg",
  description: "Fresh Grapes from Nashik.",
  created_at: "2025-05-01T00:00:00Z",
  category: "fruits",
  is_seasonal: true,
};

describe("ProductCard Snapshot", () => {
  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(<ProductCard product={sampleProduct} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
