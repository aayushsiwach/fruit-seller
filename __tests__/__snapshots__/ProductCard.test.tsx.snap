// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Snapshot renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1dc3bjb-MuiPaper-root-MuiCard-root"
  >
    <img
      alt="Grapes"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-6mo23z-MuiCardMedia-root"
      height="160"
      src="/grapes.jpg"
    />
    <div
      class="MuiCardContent-root css-2r3irp-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-69i1ev"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-ddygbf-MuiTypography-root"
          title="Grapes"
        >
          Grapes
        </h6>
      </div>
      <div
        class="MuiBox-root css-ohugnc"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1d963ga-MuiTypography-root"
        >
          Rs. 90.00
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body2 css-13rm4i2-MuiTypography-root"
        >
          Rs. 100.00
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1sbvnan-MuiTypography-root"
        title="Fresh Grapes from Nashik."
      >
        Fresh Grapes from Nashik.
      </p>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1vua6p0-MuiCardActions-root"
    >
      <div
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="9"
              cy="21"
              r="1"
            />
            <circle
              cx="20"
              cy="21"
              r="1"
            />
            <path
              d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
            />
          </svg>
        </span>
        Add to Cart
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
